### PR TITLE
more features

### LINKED ISSUE #10: full json5 support
things to do:

## strings ([chapter 5](https://spec.json5.org/#strings))
- [x] support different newline characters in multi-line strings: CR, CRLF, LS, PS
  - [x] don't include them in the parsed string
- [x] ignore other "invalid" escape sequences (for example `'\A\C\/\D\C'` is the same as `'AC/DC'`
- [x] allow U+2028 and U+2029 to appear unescaped, but produce a warning, and escape them when generating

## numbers ([chapter 6](https://spec.json5.org/#numbers))
- [x] just check that everything works, including different capitalizations of hex numbers, exponents, and +/- signs

## white space ([chapter 8](https://spec.json5.org/#white-space))
- [x] ensure that all the different types of whitespace are ignored

from [chapter 10](https://spec.json5.org/#parsers):
> An implementation may set limits on the size of texts that it accepts, the maximum depth of nesting, the range and precision of numbers, and the length and character contents of strings.

this means we could remove some things (mainly what I'm thinking of is removing `BigInteger` support, and only using `int`, `long`, and `double`).

finally we need to make sure that the writer somewhat closely follows the [grammar](https://spec.json5.org/#grammar).

### PR DESCRIPTION
**Pull request by rhysdh540**
nested objects/arrays/collections as fields in POJOs work
and for the most part (as far as I can tell) everything follows the json5 spec
also the java 8 jar is actually java 5 (which makes it smaller)
I added some more comments too, as well as publishing a javadoc jar

note that 2d arrays do *not* work, I still need to figure out how to do those nicely

closes #10

### CODE REVIEW

File: src/main/java/dev/nolij/zson/Zson.java [hunk 1]
```diff
@@ -956,7 +1038,7 @@ public void write(@NotNull Map<String, ZsonValue> data, @NotNull Appendable outp
 
 	/**
 	 * Checks if the given string is a valid identifier.
-	 * @param key The string to check.
+	 * @param key The string to check.sadasdf
```
Comment (Nolij) [inline src/main/java/dev/nolij/zson/Zson.java]: I have to assume this was left in unintentionally

**Review by Nolij — CHANGES_REQUESTED** [review]
Not a full review. Just the initial skimming of the diff.

File: src/main/java/dev/nolij/zson/Zson.java (hunk 1 above)
Comment (rhysdh540) [reply src/main/java/dev/nolij/zson/Zson.java]: yes 💀

Comment (rhysdh540) [reply src/main/java/dev/nolij/zson/Zson.java]: I needed changes in the javadoc to test whether or not doclint was turned off, so I was just adding and removing random stuff

File: build.gradle.kts [hunk 2]
```diff
@@ -114,29 +160,26 @@ val downgradeJar17 = tasks.register<DowngradeJar>("downgradeJar17") {
 }
 
 tasks.jar {
-    isPreserveFileTimestamps = false
-    isReproducibleFileOrder = true
```
Comment (Nolij) [inline build.gradle.kts:118]: Why remove these settings?

File: src/main/java/dev/nolij/zson/Zson.java [hunk 3]
```diff
@@ -993,35 +1083,52 @@ private String value(Object value) {
...
+			String indent = expandArrays ? this.indent : " ";
+			output.append(indent);
+
+			boolean first = true;
 
 			for (Object obj : iterableValue) {
-				if (expandArrays)
+				if (!first) {
+					output.append(",")
+							.append(indent);
```
Comment (Nolij) [inline src/main/java/dev/nolij/zson/Zson.java]: Weird formatting

**Review by Nolij — CHANGES_REQUESTED** [review]

File: build.gradle.kts (hunk 2 above)
Comment (rhysdh540) [reply build.gradle.kts:118]: it was moved down to a block later `tasks.withType<AbstractArchiveTask>`
not really necessary but good gradle practice

File: src/main/java/dev/nolij/zson/ZsonValue.java [hunk 4]
```diff
@@ -10,9 +11,14 @@
 public final class ZsonValue {
 
 	/**
-	 * The comment for this value. If null, there is no comment.
+	 * The value for {@link #comment} when there is no comment, represented as the null character.
 	 */
-	@Nullable
+	public static final String NO_COMMENT = "\0";
```
Comment (Nolij) [inline src/main/java/dev/nolij/zson/ZsonValue.java:16]: Does Java allow this? I feel like this might just end up as an empty string...

Comment (rhysdh540) [reply src/main/java/dev/nolij/zson/ZsonValue.java:16]: technically `"\0".length() == 1` and `"\0".isEmpty() == false`

File: src/main/java/dev/nolij/zson/Zson.java [hunk 5]
```diff
@@ -230,12 +233,32 @@ public static Map<String, ZsonValue> obj2Map(@Nullable Object object) {
...
+				} else if(value instanceof Iterable) {
+					List<Object> list = new ArrayList<>();
+					for (Object o : (Iterable<?>) value) {
+						list.add(o);
+					}
+					value = list;
+				} else if(value instanceof Map) {
+					value = obj2Map(value);
+				}
+				map.put(field.getName(), new ZsonValue(comment, value));
```
Comment (embeddedt) [inline src/main/java/dev/nolij/zson/Zson.java]: You could remove one of the `map.put` calls and save a couple more bytes. :wink:
Indentation is not correct but this gets my point across.

```java
    if(value != null) {
        if(value is array) value = new thing;
        else if(value instanceof Map) value = obj2Map(value);
    }
    map.put(field.getName(), new ZsonValue(comment, value));
```

File: src/test/java/NoJUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import dev.nolij.zson.Zson;
+import dev.nolij.zson.ZsonValue;
+
+import java.util.Map;
+
+import static dev.nolij.zson.Zson.*;
+
+public final class NoJUnitTest {
```
Comment (embeddedt) [inline src/test/java/NoJUnitTest.java]: Why do we need this?

**Review by embeddedt — CHANGES_REQUESTED** [review]

File: src/test/java/NoJUnitTest.java (hunk 6 above)
Comment (rhysdh540) [reply src/test/java/NoJUnitTest.java]: honestly I kind of forgot but I wanted to test some stuff outside of gradle
it could be removed

Comment (embeddedt) [reply src/test/java/NoJUnitTest.java]: Let's convert any new tests in it to JUnit.

File: src/main/java/dev/nolij/zson/Zson.java (hunk 5 above)
Comment (rhysdh540) [reply src/main/java/dev/nolij/zson/Zson.java]: Good catch, that would've put null value in the map and then immediately overwritten it

**Review by Nolij — APPROVED** [review]

**Pull request merged**